src/algebra/module_regular.lean

```diff
@@ -0,0 +1,167 @@
+⟨is_regular.of_smul _, ha.smul⟩
+
+end has_scalar
+
+section monoid
+
+variables [monoid R] [mul_action R M]
+
+/--  Two elements `a` and `b` are `M`-regular if and only if both products `a • b` and `b • a`
+are `M`-regular. -/
+lemma is_regular_mul_and_mul_iff :
+  is_regular M (a * b) ∧ is_regular M (b * a) ↔ is_regular M a ∧ is_regular M b :=
+begin
+  refine ⟨_, _⟩,
+  { rintros ⟨ab, ba⟩,
+    exact ⟨ba.of_smul b, ab.of_smul a⟩ },
+  { rintros ⟨ha, hb⟩,
+    exact ⟨(ha.smul_iff b).mpr hb, (hb.smul_iff a).mpr ha⟩ }
+end
+
+/--  The "most used" implication of `mul_and_mul_iff`, with split hypotheses, instead of `∧`. -/
+lemma is_regular.and_of_mul_of_mul (ab : is_regular M (a * b)) (ba : is_regular M (b * a)) :
+  is_regular M a ∧ is_regular M b :=
+⟨ba.of_smul b, ab.of_smul a⟩
+
+@[simp] lemma is_regular_one : is_regular M (1 : R) :=
+begin
+  intros a b ab,
+  rwa [one_smul, one_smul] at ab,
+end
+
+/--  Any power of an `M`-regular element is `M`-regular. -/
+lemma is_regular.pow (n : ℕ) (ra : is_regular M a) : is_regular M (a ^ n) :=
+begin
+  induction n with n hn,
+  { simp },
+  { exact (ra.smul_iff (a ^ n)).mpr hn }
+end
+
+/--  An element `a` is `M`-regular if and only if a positive power of `a` is `M`-regular. -/
+lemma is_regular.pow_iff {n : ℕ} (n0 : 0 < n) :
+  is_regular M (a ^ n) ↔ is_regular M a :=
+begin
+  refine ⟨_, is_regular.pow n⟩,
+  rw [← nat.succ_pred_eq_of_pos n0, pow_succ', ← smul_eq_mul],
+  exact is_regular.of_smul _,
+end
+
+end monoid
+
+section monoid_with_zero
+
+variables [monoid_with_zero R] [monoid_with_zero S] [has_zero M] [mul_action_with_zero R M]
+  [mul_action_with_zero R S] [mul_action_with_zero S M] [is_scalar_tower R S M]
+
+/--  The element `0` is `M`-regular if and only if `M` is trivial. -/
+lemma is_regular.subsingleton (h : is_regular M (0 : R)) : subsingleton M :=
+⟨λ a b, h (by repeat { rw mul_action_with_zero.zero_smul })⟩
+
+/--  The element `0` is `M`-regular if and only if `M` is trivial. -/
+lemma is_regular_zero_iff_subsingleton : is_regular M (0 : R) ↔ subsingleton M :=
+⟨λ h, h.subsingleton, λ H a b h, @subsingleton.elim _ H a b⟩
+
+/--  The `0` element is not `M`-regular, on a non-trivial semimodule. -/
+lemma not_is_regular_zero_iff : ¬ is_regular M (0 : R) ↔ nontrivial M :=
+begin
+  rw [nontrivial_iff, not_iff_comm, is_regular_zero_iff_subsingleton, subsingleton_iff],
+  push_neg,
+  exact iff.rfl
+end
+
+/--  The element `0` is `M`-regular when `M` is trivial. -/
+lemma is_regular_zero [sM : subsingleton M] : is_regular M (0 : R) :=
+is_regular_zero_iff_subsingleton.mpr sM
+
+/--  The `0` element is not `M`-regular, on a non-trivial semimodule. -/
+lemma not_is_regular_zero [nM : nontrivial M] : ¬ is_regular M (0 : R) :=
+not_is_regular_zero_iff.mpr nM
+
+/-- An element of `S` admitting a left inverse in `R` is `M`-regular. -/
+lemma is_regular_of_smul_eq_one (h : a • s = 1) : is_regular M s :=
+is_regular.of_smul a (by { rw h, exact is_regular_one })
```

Comment:
I think a copy of this for `of_mul_eq_one` would be a good idea too.

Reply:
Done!